# eccDNA_RCA_nanopore

eccDNA identification from nanopore long reads of rolling-circle amplicon

#### Dependency (python packages)

- [pyfaidx](https://pypi.org/project/pyfaidx/)
- [pyfastx](https://pypi.org/project/pyfastx/)
- [Biopython](https://biopython.org)

#### Usage

```
required arguments for input:
  --fastq <STR>         input reads in fastq format
  --paf <STR>           alignments file in PAF format generated by minimap2
  --reference <STR>     reference genome file in fasta format


required arguments for output:
  --info <STR>          output file for sequences information
  --seq <STR>           output file for consensus sequences in fasta format
  --var <STR>           output file for variants


optional arguments:
  --maxOffset <INT>     maximum offset of start/end positions between two sub-reads
                        to be considered as mapping to the same location [default: 20]
  --minMapQual <INT>    minimum mapping quality of sub-reads [default: 30]
  --minDP <INT>         minimum depth to call variants [default: 4]
  --minAF <FLOAT>       minimum alternative allele frequency to call variants [default: 0.75]

  --verbose             print details of consensus construction
  -h, --help            show help message
```

